lib/Controller/SMimeCertificatesController.php

```diff
@@ -0,0 +1,137 @@
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License, version 3,
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>
+ *
+ */
+
+namespace OCA\Mail\Controller;
+
+use Exception;
+use OCA\Mail\Db\SMimeCertificate;
+use OCA\Mail\Exception\ServiceException;
+use OCA\Mail\Service\Attachment\UploadedFile;
+use OCA\Mail\Service\SMimeService;
+use OCP\AppFramework\Controller;
+use OCP\AppFramework\Db\DoesNotExistException;
+use OCP\AppFramework\Http\JSONResponse;
+use OCP\IRequest;
+use OCP\Security\ICrypto;
+
+class SMimeCertificatesController extends Controller {
+	private ?string $userId;
+	private SMimeService $certificateService;
+	private ICrypto $crypto;
+
+	public function __construct(string $appName,
+								IRequest $request,
+								?string $userId,
+								SMimeService $certificateService,
+								ICrypto $crypto) {
+		parent::__construct($appName, $request);
+		$this->userId = $userId;
+		$this->certificateService = $certificateService;
+		$this->crypto = $crypto;
+	}
+
+	/**
+	 * @NoAdminRequired
+	 * @TrapError
+	 *
+	 * @throws ServiceException
+	 */
+	public function index(): JSONResponse {
+		$certificates = $this->certificateService->findAllCertificates($this->userId);
+		$certificates = array_map(function (SMimeCertificate $certificate) {
+			try {
+				$decryptedCertificate = $this->crypto->decrypt($certificate->getCertificate());
+			} catch (Exception $e) {
+				throw new ServiceException(
+					'Failed to decrypt certificate: ' . $e->getMessage(),
+					0,
+					$e,
+				);
+			}
+			$json = $certificate->jsonSerialize();
+			$json['info'] = $this->certificateService->parseCertificate($decryptedCertificate);
+			return $json;
+		}, $certificates);
+		return new JSONResponse($certificates);
+	}
+
+	/**
+	 * @NoAdminRequired
+	 * @TrapError
+	 *
+	 * @param int $id
+	 * @return JSONResponse
+	 *
+	 * @throws DoesNotExistException
+	 */
+	public function destroy(int $id): JSONResponse {
+		$this->certificateService->deleteCertificate($id, $this->userId);
+		return new JSONResponse();
+	}
+
+	/**
+	 * @NoAdminRequired
+	 * @TrapError
+	 *
+	 * @return JSONResponse
+	 *
+	 * @throws ServiceException
+	 */
+	public function create(): JSONResponse {
+		// TODO: What about PKCS12 certificates?
+		// They need to be decrypted by the client because they are protected by a password.
+		// We could use
+		//  - https://w3c.github.io/FileAPI/#reading-data-section
+		//  - https://github.com/digitalbazaar/forge#pkcs12
+		// to read and decrypt a PKCS12 certificate in the browser.
+		// The decrypted data could be attached to hidden file inputs and then sent to this endpoint
+		// as well.
+
+		$attachedCertificate = $this->request->getUploadedFile('certificate');
+		$attachedPrivateKey = $this->request->getUploadedFile('privateKey');
+
+		if ($attachedCertificate === null) {
+			throw new ServiceException('No certificate attached');
```

Comment:
use `\OCA\Mail\Http\JsonResponse` in the controller and facilitate the ::failure factory method with a HTTP STATUS_UNPROCESSABLE_ENTITY status.

Note the slightly changing json format. you need to access the data with response.data.data on the frontend

Reply:
We have a custom JsonRespone in Mail?! I also refactored to `JsonResponse::success()`.

Done